### DISCUSSION TITLE
Ghidra Trace Format

### DISCUSSION THREAD

**Discussion by pjsoberoi** [NationalSecurityAgency/ghidra#4051] [Debugger]
@d-millar,

Similar to this discussion on the [GADP](https://github.com/NationalSecurityAgency/ghidra/discussions/4033) protocol, I would like to request the creation of an open source Ghidra trace format. I can see a lot of ideas for tools that can use the trace format.

**Comment by d-millar** [comment]
@pjsoberoi going for the easy requests, I see.  :)

Would love to see an open-source trace format, but better programmers than I have tried and failed on this front.  Again, just for the sake of discussion, what would you want for the proposed format:

- optimized for collection speed
- optimized for trace size
- optimized for post-processing speed
- optimized for random access
- should it record only the instruction sequence
- record enough context to recreate state
- record enough timing information to recreate multithreaded state
- should it allow for custom events
- should it be reversible or quasi-reversible
- should it be optimized for differential analysis
- should it be compressed or human-readable
- should it be optimized for ease of use and customization

Obviously many of these options are mutually exclusive, and I’m not sure you could even option your way through them.  Also, there’s the whole issue of adoption - if you’re shooting for widespread adoption, the format has to at least be able to match the constraints of possible generators.  

Just my two cents.  Would be interested in ideas/discussion of the possibilities and possible use cases.

**Reply by gourryinverse** [reply]
Honestly, I don't think format is the important question.  I think an API for collecting that data modularly is step 1.

### Data Collection

First I think you need to consider what data is even possible to collect, and what are the performance implications of collecting each piece of data.  For example - capturing register contexts instruction-by-instruction is significantly slower than just capturing basic block addresses.  Capturing memory access information is even worse, depending on the granularity.

I would break it down into classes of data

1. Basic branching data - e.g. (from, to)  edge collection
2. Exhaustive branching data - e.g. (from, to) pairs with meta data about an event occurring.  Is the instruction that caused the branch a branching instruction or did it cause an exception? Did a timer fire? Etc.  Note that this means different things in the context of a emulator vs a userland-only environment (userland obviously simpler to an extent)
3. Memory access meta data.  e.g. (rwx) + (address) + (size). 
4. Memory access data.  e.g.  (rwx) + (address) + (size) + **(the data)**
5. Per-instruction meta data.  e.g. on switch from instruction 1 to instruction 2,  the value of rax changed.
6. Full on "trace the world" data - all of the above.
7. Custom hooks for some given event.

### Use Cases

I think there's a step here to have a "call for comments" of sorts to ask for what all people want to do with ghidra traces, and then comb through the list and ask what data is required to implement that list.  Come back around and denote what the performance implications of each subset of data, and determine if it's possible to generate those pieces individually - and modularly.

Here's the cases I can think of off the top of my head

1. basic coverage visualization (xyz blocks were hit / not hit / heatmap)
2. state transition map visualization (block X was jumped to from addresses ABCD and E, etc)
3. dataflow analysis - some memory address 0x10 was "interesting", visualize what areas of code accessed it, and how it affected the software
4. p-code based fuzzer - given some trace, and some set of data from 1-6, is it possible to do what is effectively concolic execution on pcode?

In addition - ask whether the "trace format" even matters.  What if I don't want to retain the information, and I just want to do something with it computationally during that moment and then throw it away?  You could skip the file write entirely and just expose a callback for a plugin to crunch the data. 

Which brings me to...

### File Format

If the above "callback" api for that data exists, any given file format generator can be written on top of it.  As a result - I wouldn't worry about the specific ghidra file format until you know what data you're going to provide - and how.

Just from experience - I've written an Intel Processor Trace emulator for a cross-architecture emulation platform, and found that for basic branch-trace information it's extraordinarily effective - but if you want to add memory access information is gets to be oppressive.  Like-wise I've also found that generating that style log (Intel PT) is both 

1. really computationally expensive on both sides of the operation.  We're talking 10%+ overhead at runtime, plus the added decode time on the other side), possibly more if it forces you into a bad execution mode (i.e. emulator vs bare metal).
2. mostly not useful in the context of a fuzzer that wants live feedback

So I think jumping to file format at this stage is premature.

For my plug (I am biased), I do think the Intel PT file format is extraordinarily impressive, very flexible, and the custom PTWrite packets can let you do a lot of customization.  Plus you can choose whether or not to even use the compression options, which can reduce decode time.

### External trace generators and importing traces

Whatever the ultimate file format is, consider external sources of trace information.  QEMU/KVM traces, Intel PIN, DrCOV, flat files of branch records - there are an infinite number of possibilities here because there's no one standard to rule them all.

I only bring this up because it would be nice to be able to have an interface to convert these things to some format ghidra can understand.  Obviously plugins like DragonDance already do some of this.  Just something to consider - someone somewhere is going to ask the question "well i have a trace from xyz tool, can i import that into ghidra?"

### Performance

OP suggests optimizing for collection speed, but not for trace size.  In my experience, the amount of data produced **is the primary factor** in dictating performance (i/o chokepoints baby!).  Tl;dr: They're the same problem in my opinion - but we could probably debate for days over this.  Have some ability to target "areas of interest" and omit trace overhead from anywhere else is really effective in reducing both overhead and log size.

In my experience in working with Intel PT - dialing in on trace size vs hook-induced-overhead is a heuristic problem, but the sweet spot become pretty obvious after a little bit of testing.  Before you go generating a 90GB trace file, consider what data in that file is **actually relevant** and just emit that data.  This goes back to having modularity for what data is actually produced.

**Reply by d-millar** [reply]
@gourryinverse Awesome feedback - much appreciated!

**Reply by gourryinverse** [reply]
@d-millar i realize my post was long, and maybe beyond the intent of @pjsoberoi 's post, but having gone in loops around trace and coverage formats for probably going on 5 years, adding "yet another format" to the pile is far less important than understanding what people are trying to do. 

Maybe more succinctly - this task / discussion should probably be around Ghidra Trace format**S** (plural).

**Reply by d-millar** [reply]
@gourryinverse Been there as well and agree completely with everything you’ve said.  That’s why, for instance, we’ve implemented access to windbg/kd TTD traces through their API, rather than importing their traces directly.  Am considering doing the same for the Intel PIN API and (if it exists - I haven’t done enough research yet) for the Intel PT API.  (Any pointers on the latter would be greatly appreciated!)

I do believe tracing is one of those areas like visualization - it has a lot of promise and is very seductive, but is easy to overestimate.  Targeted tracing has, for me at least, been much more protective in the past.  That said, I do understand PJ’s use case, and it’s quite possible none of the aforementioned technologies really solves his problem.  I do think the more opinions/insights we can get on this the better.  It’s an extremely interesting, although definitely research-required problem.

**Reply by d-millar** [reply]
I should also mention having a few more test cases, such as Intel PT or RR, would allow us to flesh out our ideas about interaction between the Ghidra “traces”, which are more of a session record in database format, and more global trace formats, particularly in the area of time-scale information and reversible debugging interfaces.

**Reply by gourryinverse** [reply]
@d-millar if you want a crash-course on Intel PT, I'd be happy to give you 20-30 minutes and do a quick walk through of what I know.   I would say it's a pretty well documented binary stream, the issue with it is that if you have *real* Intel PT data (produced by the chip, rather than from an emulated system or such), you might not necessarily be able to "jump to" any particular part of the trace and make sense of it, due to 1) Variable packet length, and 2) a variety of compression cheats they use that make decoding particularly expensive.

**Reply by d-millar** [reply]
@gourryinverse That would be extremely useful!  Am traveling this week, but am definitely interested.  Will give you a shout when I get back.

**Comment by pjsoberoi** [comment]
Obviously you have spent much more time researching this area than I have. I have not considered possible format optimizations at all. 
One use case I'm interested in is processor and processor module verification. For that I would need all instructions, all registers, and memory deltas. 

The first idea that came to mind is to maybe have a simple raw trace format that can be optimized post collection into another format. The goals would be to optimize for collection speed, simple format to implement (for reading and writing), and tools to convert to other optimized formats for different uses cases. 

- optimized for collection speed

Yes. 

- optimized for trace size

No. Another program can optimize the trace afterwards into another format. 

- optimized for post-processing speed

No. Another program can optimize the trace afterwards into another format. 

- optimized for random access

No. Another program can optimize the trace afterwards into another format. 

- should it record only the instruction sequence

Maybe have flags for instructions, instructions + regs, instructions + regs + memory deltas. I can do useful stuff with just instructions + regs, I don't know what I could do with just instructions. 

- record enough context to recreate state

All state encapsulated by instructions + regs + memory deltas seems doable. State such as other processes, kernel, network stack, etc seem really hard. I would skip that for now. 

- record enough timing information to recreate multithreaded state

Sure. But that can be an option. Seems hard so skip it for now. 

- should it allow for custom events

No. Another program can optimize the trace afterwards into another format and that can handle custom events. 

- should it be reversible or quasi-reversible

Yes. 

- should it be optimized for differential analysis

No. Another program can optimize the trace afterwards into another format. 

- should it be compressed or human-readable

No. Another program can take the trace as input and transform it to something human readable. 

- should it be optimized for ease of use and customization

Yes.

**Comment by d-millar** [comment]
A bit random, but wanted to say I haven't forgotten about this - just haven't had time to really take it on.

**Comment by nsadeveloper789** [comment]
Related: https://github.com/NationalSecurityAgency/ghidra/discussions/2730#discussioncomment-10424795
